Fix checkboxes in Firefox and IE/Edge

patrickkettner opened the pull request:
firefox doesn't allow for the input elements style to be completely overwritten
like it has been in chrome. So instead I am hiding the checkbox via opacity and
loading the checkbox as the background of the label. This has the nice side
effect of fixing an issue with Edge and svg backgrounds on input elements.
Finally, firefox is choking on `#` in svg data uris. escaping the url fixes
that.

closes tastejs/todomvc#1408

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/tastejs/todomvc-app-css/8)

<!-- Reviewable:end -->

sindresorhus commented on index.css:
```diff
@@ -117,28 +117,32 @@ input[type="checkbox"] {
 }
 
-.toggle-all {
+.toggle-all + label{
```
```
.toggle-all + label {
```

sindresorhus commented:
> Finally, firefox is choking on # in svg data uris. escaping the url fixes
that.

I'd prefer we not do this. It makes the SVG totally unreadable. Is there a Firefox bug ticket for this?

sindresorhus commented:
Nice work @patrickkettner! :)

samccone commented:
Also will close https://github.com/tastejs/todomvc/issues/1259

thanks a million @patrickkettner :+1: :+1: :+1: :+1: :+1:  :rocket:

patrickkettner commented:
@sindresorhus just escaping the `#` works too. I reduced it to just that bit. Not 100% intuitive (should inline comments be added?). Couldn't find a relevant FF bug, but I planned on opening one this afternoon.

passy commented:
@patrickkettner This looks great! But it seems that the `toggle-all` checkbox is no longer clickable after the change.

patrickkettner commented:
@passy hmmm... it is for me. What browser are and framework are you testing it in?

sindresorhus commented:
ping @passy

samccone commented:
pong @passy :wink:

passy commented:
@patrickkettner 

Browsers tested:

- Chrome 44.0.2403.130 on Linux
- Firefox 41.0a2 (2015-08-03) on Linux
- Chrome 45.0.2454.62 on Android
- Safari 8.0.8 on OS X

patrickkettner commented:
Pardon?

On Sat, Aug 22, 2015, 5:19 PM Pascal Hartig <notifications@github.com>
wrote:

> @patrickkettner <https://github.com/patrickkettner>
>
> Browsers tested:
>
>    - Chrome 44.0.2403.130 on Linux
>    - Firefox 41.0a2 (2015-08-03) on Linux
>    - Chrome 45.0.2454.62 on Android
>    - Safari 8.0.8 on OS X
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tastejs/todomvc-app-css/pull/8#issuecomment-133755439>
> .
>

passy commented:
Sorry, these are all the browsers I've tried and managed to reproduce it in.

patrickkettner commented:
no worries, I actually tried them on all of browserstacks desktop browsers and didn't spot any issues. Haven't tried mobile, however

arthurvr commented:
I am able to repro the same issues as @passy on Chrome 46 on OS X.

patrickkettner commented:
thanks, @arthurvr. ill check it out shortly

patrickkettner commented:
which projects are you testing this against?

On Sun, Aug 23, 2015 at 2:22 PM, Arthur Verschaeve <notifications@github.com
> wrote:

> I am able to repro the same issues as @passy <https://github.com/passy>
> on Chrome 46 on OS X.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tastejs/todomvc-app-css/pull/8#issuecomment-133892461>
> .
>



-- 
patrick

sindresorhus commented:
ping @arthurvr @passy

passy commented:
@patrickkettner I think I used `typescript-backbone` because it uses the new CSS structure.

patrickkettner commented:
So the reason it is broken on that particular one is because the label/input connection is broken. Its using a `for` attribute that is not associated with any `id` or `name`.  I would view this as a bug in the html, but I can make adjustments on the css end if y'all think this is the right thing to do

patrickkettner commented:
@passy @sindresorhus So the reason it is broken on that particular one is because the label/input connection is broken. Its using a for attribute that is not associated with any id or name. I would view this as a bug in the html, but I can make adjustments on the css end if y'all think this is the right thing to do

passy commented:
I'm pretty sure I've tried three in total. If an HTML change is required I
think that's alright. Which app does it work with unmodified so I can try
it as it is?

On Thu, Sep 10, 2015, 10:40 patrick kettner <notifications@github.com>
wrote:

> @passy <https://github.com/passy> @sindresorhus
> <https://github.com/sindresorhus> So the reason it is broken on that
> particular one is because the label/input connection is broken. Its using a
> for attribute that is not associated with any id or name. I would view this
> as a bug in the html, but I can make adjustments on the css end if y'all
> think this is the right thing to do
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tastejs/todomvc-app-css/pull/8#issuecomment-139168931>
> .
>

samccone commented:
>  it is broken on that particular one is because the label/input connection is broken

^^ makes sense, imho we can just manually go in and introduce a line or two of CSS to resolve this per framework

patrickkettner commented:
@passy I am not sure where I was testing it before, but if you add `id=toggle-all` or `name=toggle-all` to the markup on the typescript-backbone example you will see it working

@samccone just to clarify, this is something that should be fixed in HTML, rather than CSS (in my opinion, that is)

samccone commented:
Unfortunately some of the example frameworks have limitations around the DOM structure, but we will just need to enumerate all of the broken ones and list out how we will fix them.

patrickkettner commented:
there are frameworks that don't allow for an id?

patrickkettner commented:
Ping @samccone @passy

sindresorhus commented:
@patrickkettner Yes, let's try to get this merged, finally.

> just escaping the # works too. I reduced it to just that bit. Not 100% intuitive (should inline comments be added?). Couldn't find a relevant FF bug, but I planned on opening one this afternoon.

There needs to be a link to an existing or new FF bug ticket as an inline comment.

---

I tried this change with the Backbone app and compared with existing CSS and noticed a few incorrect changes:

- The toggle all button moved. Make sure it's in the exact same spot.
- Long lines are no longer wrapped.  <img width="746" alt="screen shot 2017-01-24 at 11 59 01" src="https://cloud.githubusercontent.com/assets/170270/22234723/e7636602-e22c-11e6-9f84-175c75e21bad.png">

---

Someone needs to test this in IE and latest Edge and confirm it's working fine there. // @jdalton

---

As for the broken toggle-all connection, we can just make this a major version in this package, and ensure we also update https://github.com/tastejs/todomvc-app-template/blob/master/index.html (Submitting a PR there with a fix would help move this along)

patrickkettner commented:
@sindresorhus thank you for taking the time to look at this

so, ironically, reducing the escaped value for Firefox broken it for Edge and IE. They _require_ the full value to be escaped. There is an [open bug](https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/7157459/) to fix this, which I have marked as highly important internally due to todo's reliance (amongst others) on it. As a result I reverted the reduced escaping back to the full escape.
I also created a firefox bug for the hash issue, which is also commented on in the css.

The long lines were a type on my part ("pre" !== "pre-wrap"), which was later changed in master by using `word-break: break-all`. It is now also fixed

I fixed the placement of the toggle all button by tweaking the absolute positioning.

As for the toggle-all not being clickable, I am not able to reproduce it. If it _is_ reproducible anywhere, it would be due to that demo having broken html (which I would be happy to PR a fix for, if anyone finds it).

Let me know if there is anything else I can do to get this merged!

patrickkettner commented:
ping @samccone @sindresorhus

sindresorhus commented:
I tried your changes with the [Backbone example](http://todomvc.com/examples/backbone/) and the template and now the text goes over the checkbox.

<img width="181" alt="screen shot 2017-03-15 at 13 54 44" src="https://cloud.githubusercontent.com/assets/170270/23936977/9372b21e-0987-11e7-8e6f-b061c678f999.png">

Can you please test your changes against the template or the Backbone example? I say this in the kindest way possible, but I could not care less about this change and there are so many other things I rather do than go back-and-forth on this PR :)

patrickkettner commented:
apologies - that did not occur in `typescript-backbone`. updated accordingly

sindresorhus commented:
Thank you! Glad to see this finally merged :)

patrickkettner commented:
🎉 🎉 🎉
